### PR TITLE
Add Olmo3 model

### PR DESCRIPTION
**Pull request by 2015aroras**
# What does this PR do?

This PR adds the implementation for the upcoming Olmo 3 model. The main architectural differences from Olmo 2 are:
- Sliding window attention is used for 3 out of 4 layers. RoPE scaling is not applied to sliding window attention layers.

It's possible to add sliding window functionality directly into the Olmo 2 code, so if it is preferred then I can implement it that way.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Having a look!

**Comment by 2015aroras** [comment]
FYI, vLLM prefers having the Olmo3 changes added to their Olmo2 implementation rather than put in a separate implementation (so that's how I am implementing Olmo3 there).

File: src/transformers/models/olmo3/modular_olmo3.py [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+        """
+        rope_config_validation(self)
+
+
+class Olmo3RMSNorm(Olmo2RMSNorm):
+    pass
+
+
+# Olmo3 attention is identical to OLMo 2 attention except:
+# - Sliding window attention is used for 3 out of 4 layers.
```
Comment (ArthurZucker) [inline src/transformers/models/olmo3/modular_olmo3.py:240]: nice!

Comment (ArthurZucker) [reply src/transformers/models/olmo3/modular_olmo3.py:240]: I think we can use `Exaone4Attention` here!

File: src/transformers/models/olmo3/modular_olmo3.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+class Olmo3Attention(Olmo2Attention):
```
Comment (ArthurZucker) [inline src/transformers/models/olmo3/modular_olmo3.py:241]:
```suggestion
class Olmo3Attention(Exaone4Attention):
```

File: src/transformers/models/olmo3/modular_olmo3.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+            # Create the masks
+            causal_mask_mapping = {
+                "full_attention": create_causal_mask(**mask_kwargs),
+                "sliding_attention": create_sliding_window_causal_mask(**mask_kwargs),
+            }
+
+        hidden_states = inputs_embeds
+        position_embeddings_mapping = {
+            layer_type: rotary_emb(hidden_states, position_ids) for layer_type, rotary_emb in self.rotary_embs.items()
+        }
```
Comment (ArthurZucker) [inline src/transformers/models/olmo3/modular_olmo3.py]:
```suggestion
        position_embeddings_mapping = {
            "sliding_attention": self.rotary_embs["sliding_attention"](hidden_states, position_ids),
            "full_attention": self.rotary_embs["full_attention"](hidden_states, position_ids),
        }
```
we'd much rather have something explicit!

File: src/transformers/models/olmo3/modular_olmo3.py [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+    def __init__(self, config: Olmo3Config):
+        super().__init__(config)
+        self.norm = Olmo3RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        self.layers = nn.ModuleList(
+            [Olmo3DecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
+        )
+        self.rotary_embs = {
+            "sliding_attention": Olmo3RotaryEmbedding(config=config, rope_type="default"),
+            "full_attention": Olmo3RotaryEmbedding(config=config),
+        }
```
Comment (ArthurZucker) [inline src/transformers/models/olmo3/modular_olmo3.py]:
```suggestion
        self.rotary_embs = nn.ModuleDict({
            "sliding_attention": Olmo3RotaryEmbedding(config=config, rope_type="default"),
            "full_attention": Olmo3RotaryEmbedding(config=config),
        })
```
I just want to make sure it works well with auto device placements!

File: src/transformers/models/olmo3/modular_olmo3.py [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+
+        self.max_seq_len_cached = config.max_position_embeddings
+        self.original_max_seq_len = config.max_position_embeddings
+
+        self.config = config
+        self.rope_init_fn = ROPE_INIT_FUNCTIONS[self.rope_type]
+
+        inv_freq, self.attention_scaling = self.rope_init_fn(self.config, device)
+        self.register_buffer("inv_freq", inv_freq, persistent=False)
+        self.original_inv_freq = self.inv_freq
```
Comment (ArthurZucker) [inline src/transformers/models/olmo3/modular_olmo3.py:323]: I think we can just use `llama`

**Review by ArthurZucker — APPROVED** [review]
Very nice thanks! 🤗

File: src/transformers/models/olmo3/modular_olmo3.py (hunk 1 above)
Comment (2015aroras) [reply src/transformers/models/olmo3/modular_olmo3.py:240]: Is there a benefit to using `Exaone4Attention` as the parent class here? Exaone4 only applies rope only sliding window layers whereas Olmo3 applies rope to all layers, so `forward()` would need to be overriden anyways. Also Exaone4 uses head-wise qk norm, which would need to be undone too. So it seems like Olmo3 attention is closer to Olmo2 attention than Exaone4 attention.

File: src/transformers/models/olmo3/modular_olmo3.py (hunk 5 above)
Comment (2015aroras) [reply src/transformers/models/olmo3/modular_olmo3.py:323]: RoPE scaling is not applied to sliding window attention layers in Olmo3. To support this, I allow passing in the rope type in the `Olmo3RotaryEmbedding` constructor and add lines 306-7 to use the passed in type. Thus I can't fully reuse the existing rope implementations like that of Olmo2 or Llama.

File: src/transformers/models/olmo3/modular_olmo3.py (hunk 4 above)
Comment (2015aroras) [reply src/transformers/models/olmo3/modular_olmo3.py]: https://github.com/huggingface/transformers/pull/40778/commits/7cd9657751c23d635bd1ee74d3bbcce44055bf9e Done

File: src/transformers/models/olmo3/modular_olmo3.py (hunk 3 above)
Comment (2015aroras) [reply src/transformers/models/olmo3/modular_olmo3.py]: https://github.com/huggingface/transformers/pull/40778/commits/7cd9657751c23d635bd1ee74d3bbcce44055bf9e Done

File: src/transformers/models/olmo3/modular_olmo3.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/olmo3/modular_olmo3.py:240]: okay no worries I checked fast so not 100% sure odn't worry!

File: src/transformers/models/olmo3/modular_olmo3.py (hunk 5 above)
Comment (ArthurZucker) [reply src/transformers/models/olmo3/modular_olmo3.py:323]: Okay no worries, I thought the llama one was more permisive to this!

**Comment by ArthurZucker** [comment]
run-slow: auto, olmo3

**Review by ArthurZucker — APPROVED** [review]
Nice! Thanks for the PR! 🤗 happy to merge!

**Comment by 2015aroras** [comment]
Pulling main modified the 7b logits slightly somehow 🤷. Updated that test, but pulling main has also led to irrelevant test failing.

**Comment by ArthurZucker** [comment]
Now worries unrelated can be skipped! Merging 🤗 thanks a lot for the PR !

**Comment by 2015aroras** [comment]
Thank you for the quick reviews!
